File: 338. Counting Bits/solution.cpp
/*
3 ms
7.9 MB
*/

/**
 * TC: O(n)
 * SC: O(1)
*/

/**
 * @author : SahilK-027
 * @brief : Dynamic Programming
*/

class Solution {
public:
    vector<int> countBits(int n) {
        // Build dp
        vector<int> dp(n+1, 0);
        // Initial round will be 1 
        // We will be changing round at [1,2,4,8,16,32,64...]
        int curr_round = 1;
        for(int i = 1 ; i <= n ; i++){
            if(curr_round * 2 == i){
                curr_round = i;
            }
            dp[i] = dp[i - curr_round] + 1;
        }
        return dp;
    }
};

File: 338. Counting Bits/solution.js
/*
86 ms
54.1MB
*/

/**
 * TC: O(n log n)
 * SC: O(1)
*/

/**
 * @author : SahilK-027
 * @brief : iterative
*/
/**
 * @param {number} n
 * @return {number[]}
 */
var countBits = function(n) {
    ans = [];
    for(let i = 0 ;i <= n; i++){
        let curr_num = i;
        let cnt =0;
        while(curr_num){
            let remender = curr_num % 2;
            if(remender == 1) cnt++;
            curr_num = parseInt(curr_num /2);
        }
        ans.push(cnt);
    }
    return ans;
};